server/src/instant/util/aws_signature.clj

```diff
@@ -0,0 +1,352 @@
+(ns instant.util.aws-signature
+  "Implements AWS' Signature Version 4. [1] 
+   
+   We use these to generate S3 presigned URLs.
+   
+   Why not use the Java SDK directly to generate presigned URLs? 
+     We need a way to set the signing time. 
+     A consistent signing-time gives us stable URLs, which browsers can cache. 
+   
+   This code was originally imported from clj-aws-sign [2]
+
+   [1] https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_sigv-create-signed-request.html
+   [2] https://github.com/joseferben/clj-aws-sign/blob/master/src/clj_aws_sign/core.clj"
+  (:require
+   [clojure.string :as str]
+   [instant.util.crypt :as crypt-util])
+  (:import
+   (java.net URLEncoder)
+   (java.time ZonedDateTime ZoneId Instant Duration)
+   (java.time.format DateTimeFormatter)))
+
+(set! *warn-on-reflection* true)
+
+;; ------------
+;; Constants 
+
+(def sig-algorithm "AWS4-HMAC-SHA256")
+
+(def unsigned-payload "UNSIGNED-PAYLOAD")
+
+(def empty-sha256 "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")
+
+;; ------------------------
+;; CanonicalRequest 
+
+(defn- url-encode
+  "Percent encode the string to put in a URL."
+  [^String s]
+  (-> s
+      (URLEncoder/encode "UTF-8")
+      (.replace "+" "%20")
+      (.replace "*" "%2A")
+      (.replace "%7E" "~")))
+
+(defn- change-directory
+  [segments]
+  (let [change-amount (get (frequencies segments) ".." 0)]
+    (if (> change-amount 0)
+      (change-directory (drop-last (rest segments)))
+      segments)))
+
+(defn- resolve-path
+  [path]
+  (->> (str/split path #"/")
+       (remove
+        (fn [s]
+          (or (str/blank? s)
+              (= "." s))))
+       (change-directory)
+       (str/join "/")
+       (str "/")))
+
+(defn- encode-path
+  [path]
+  (->> (str/split path #"/")
+       (map url-encode)
+       (str/join "/")
+       (#(if (str/blank? %) "/" %))))
+
+(defn- append-slash
+  [path raw]
+  (str path (and (re-matches #".*/$" raw) "/")))
```

Comment:
A comment here would be helpful. It seems like it's only appending a slash if it already had a slash? If so, would it be more efficient to do

`(when (.endsWith raw "/") "/")`